Methods label functions (#6)
greenelab/text_mined_hetnet_manuscript: content/03.methods.md

Talk about preprocessing Pubtator
Talk about hand annotations for each realtion

## Label Functions for Annotating Sentences

> We used simple pythonic functions called label functions to quickly annotate our candidates sentences.

Reviewer: Start by bringing your reader into the idea of label functions in the context of the snorkel-style approach.
1-2 sentences introducing the broad idea of Snorkel's approach.
1 sentence on conceptually what label functions do within that context.
1 sentence on conceptually how label functions are combined (note the generative/discriminative model sections coming later).

Then start.

> These functions are designed to emit: a positive label (1), a negative label (-1) 

Reviewer: ```suggestion
These functions emit: a positive label (1), a negative label (-1) 
```

(also note, one-sentence per line was broken here)

or abstain from emitting a label (0).

> We constructed label functions in the following three categories (Databases, Text Patterns and Domain Heuristics).

Reviewer: ```suggestion
We constructed three categories of label functions: databases, text patterns and domain heuristics.
```

These categories are described below using the following candidate sentence as an example: \"<span style="color:blue">PTK6</span> may be a novel therapeutic target for <span style="color:brown">pancreatic cancer</span>."

> **Databases**: This category of label functions are designed to utilize publicly available databases to determine if a co-mention pair has been previously established.

Reviewer: ```suggestion
**Databases**: Label functions in this category return a result based on whether or not a relationship has been established between the co-mention pairs in a certain database.
```

> It emits a positive label if the co-mention pair is located within a database, otherwise the given label function will abstain.

Reviewer: ```suggestion
They emit a positive label if a relationship between the co-mentioned pair exists within the database.
If there is no relationship, they abstain.
```

> If the label function abstains, another label function will emit a negative label.

Reviewer: I don't quite get what this is going at without the context of our past conversations - re-word

$$ \Lambda_{DB}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 1 & (\color{#875442}{D}, \color{#02b3e4}{G}) \in DB \\
0 & otherwise \\
\end{cases} $$

$$ \Lambda_{\neg DB}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 -1 & (\color{#875442}{D}, \color{#02b3e4}{G}) \notin DB \\
0 & otherwise \\
\end{cases} $$

> **Text Patterns**: This category of label functions are designed to provide contextual information such as keywords, number of words between two mentions and rules on grammatical structure of a sentence.

Reviewer: ```suggestion
**Text Patterns**: These label functions are designed to provide information about the relationship of the entities in the sentence such as whether keywords are found in the sentence, the number of words between two mentions and rules on grammatical structure of a sentence.
```
Since the other things are specific examples, can you give a specific example on the grammatical rules?

> The polarity of these functions can vary between positive and negative depending on the situation.

Reviewer: ```suggestion
These functions emit a positive or negative output depending on the situation.
In general, those focused on keywords emit positives and those focused on negation emit negatives.
```

Reviewer: I think the second bit is true, but if it's not it should be adjusted. I thought something like this would help the reader grasp why it could be either.

$$ \Lambda_{TP}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 1 & "target" \in Candidate \> Sentence \\
 0 & otherwise \\
\end{cases} $$

> **Domain Heuristics**: This category of label functions is designed to incorporate domain expert knowledge and heuristic rules to annotation sentences.

Reviewer: ```suggestion
**Domain Heuristics**: This category of label functions uses domain-specific heuristic rules to annotate sentences.
```

> We used the findings of Percha et al.[@doi:10.1093/bioinformatics/bty114] to generate label functions of this category.

Reviewer: ```suggestion
We used Percha et al.[@doi:10.1093/bioinformatics/bty114] as this category's source of label functions.
```

> If a sentence was highlighted in their study, our label functions emitted a positive label.

Reviewer: I don't have a specific suggestion here, but I don't know what highlighted means so this probably needs to be re-worded.

$$
\Lambda_{DH}(\color{#875442}{D}, \color{#02b3e4}{G}) = \begin{cases}

>     1 & Candidate \> Sentence \in Experimental \> Results\\

Reviewer: Is `Experimental Results` a specific thing in their work or your term for it? Make sure your terminology matches what they use in their paper :)

    0 & otherwise \\
    \end{cases}
$$

> After constructing these label functions we ended up with the following distribution (shown in Table 2).

Reviewer: ```suggestion
Roughly half of our label functions are based on text patterns, while the others are distributed across the databases and domain heuristics (Table 2).
```

| Relationship | Databases (DB) | Text Patterns (TP) | Domain Heuristics (DH) |
| --- | :---: | :---: | :---: |
| Disease associates Gene (DaG) | 7 | 20 | 10 | 
| Compound treats Disease (CtD) | 3 | 15 | 7 |
| Compound binds Gene (CbG) | 9 | 13 | 7 | 
| Gene interacts Gene (GiG) | 9 | 20 | 8 | 

Table 2. The distribution of each label function per relationship. 

> Some relationships take more label functions than others to achieve acceptable performance.

Reviewer: ```suggestion
```
Probably remove this here, since it gets into the results. Plus nobody wants to fight with a reviewer over the definition of acceptable performance. :)

Talk about preprocessing Pubtator
Talk about hand annotations for each realtion

## Label Functions for Annotating Sentences

> In natural language processing a common problem is having too much data and not enough ground truth annotations.

Reviewer: ```suggestion
A common challenge in natural language processing is having too few ground truth annotations, even when data are abundant.
```

> A common challenge in natural language processing is having too few ground truth annotations, even when data are abundant.

Reviewer 2: possible to replace "data" with something more specific... are you referring to "input text" or the like?

> Data programming [@arxiv:1605.07723] circumvents this issue by quickly annotating large datasets by using multiple noisy signals emitted by label functions.

Reviewer 2: I'd place the reference at the end of this sentence because the whole sentence describes data programming. In general references usually go at the end of the setnence, unless additional clarity on what the citation applies to is provided by interspersing them.

> Label functions are simple pythonic functions emit: a positive label (1), a negative label (-1) or abstain from emitting a label (0).

Reviewer 2: Grammer:

> Label functions are simple pythonic functions emit

Do you mean "Pythonic" as in adhering to the ideal python idioms? Or do you just mean are written in Python?

Author: Could be both, but for our case these are functions written in python. Technically, label functions could be written in any desired language. Python is one of the easiest to deal with.

> We combine these functions using a generative model to output a single annotation, then use a discriminator model to improve upon this model.

Reviewer 2: Can we expound on what "annotation" means? It is essential the reader understands that the annotation is a consensus score among labeling fxns that ranges from 0 to 1 (i.e. a consensus outcome probability).

> Label functions fall into these three categories: databases, text patterns and domain heuristics.

Reviewer: ```suggestion
Our label functions fall into these three categories: databases, text patterns and domain heuristics.
```

> Our label functions fall into these three categories: databases, text patterns and domain heuristics.

Reviewer 2: "these" can be removed

> These categories are described below using the following candidate sentence as an example: \"<span style="color:#02b3e4">PTK6</span> may be a novel therapeutic target for <span style="color:#875442">pancreatic cancer</span>."

Reviewer: ```suggestion
We provide examples for the categories, described below, using the following candidate sentence: \"<span style="color:#02b3e4">PTK6</span> may be a novel therapeutic target for <span style="color:#875442">pancreatic cancer</span>."
```

> We provide examples for the categories, described below, using the following candidate sentence: \"<span style="color:#02b3e4">PTK6</span> may be a novel therapeutic target for <span style="color:#875442">pancreatic cancer</span>."

Reviewer 2: Talk to @vincerubinetti about how to do colors as per https://manubot.github.io/rootstock/v/ed4158c9f2a1b300f82d058be9aefe1f5eb72763/#special

**Databases**: These label functions incorporate existing databases to generate a signal, as seen in distant supervision [@doi:10.3115/1690219.1690287].

> These functions assert if a candidate sentence's co-mention pair is present in a given database.

Reviewer 2: check/detect/query is more accurate than assert?

If the pair is present, emit a positive label and abstain otherwise.
If the pair isn't present in any existing database, then a separate label function will emit a negative label.

> We use a separate label function to prevent the label imbalance problem, which is less studied candidates will be drowned out by negative labels.

Reviewer 2: Rephrase?
> which is when less studied candidates are drowned out by negative labels.

These negative labels increase the likelihood of misclassification when training the generative model.

$$ \Lambda_{DB}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 1 & (\color{#875442}{D}, \color{#02b3e4}{G}) \in DB \\
0 & otherwise \\
\end{cases} $$

$$ \Lambda_{\neg DB}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 -1 & (\color{#875442}{D}, \color{#02b3e4}{G}) \notin DB \\
0 & otherwise \\
\end{cases} $$

**Text Patterns**: These label functions are designed to use keywords and sentence context to generate a signal. 
For example, a label function could focus on the number of words between two mentions or focus on the grammatical structure of a sentence.
These functions emit a positive or negative output depending on the situation.
In general, those focused on keywords emit positives and those focused on negation emit negatives.

$$ \Lambda_{TP}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}
 1 & "target" \> \in Candidate \> Sentence \\
 0 & otherwise \\
\end{cases} $$

$$ \Lambda_{TP}(\color{#875442}{D}, \color{#02b3e4}{G}) = 
\begin{cases}

>  -1 & 	`"VB" \> \notin pos\_tags(Candidate \> Sentence) \\

Reviewer: Does this backtick "`" ever end? The syntax highlighting in github says it doesn't

 0 & otherwise \\
\end{cases} $$

**Domain Heuristics**: These label functions use the other experiment results to generate a signal. 
For this category, we used dependency path cluster themes generated by Percha et al [@doi:10.1093/bioinformatics/bty114].
If a candidate sentence's dependency path belongs to a previously generated cluster, then the label function will emit a positive label and abstain otherwise.

$$
\Lambda_{DH}(\color{#875442}{D}, \color{#02b3e4}{G}) = \begin{cases}
    1 & Candidate \> Sentence \in Cluster \> Theme\\
    0 & otherwise \\
    \end{cases}
$$

Roughly half of our label functions are based on text patterns, while the others are distributed across the databases and domain heuristics (Table 2).

| Relationship | Databases (DB) | Text Patterns (TP) | Domain Heuristics (DH) |
| --- | :---: | :---: | :---: |
| Disease associates Gene (DaG) | 7 | 20 | 10 | 
| Compound treats Disease (CtD) | 3 | 15 | 7 |
| Compound binds Gene (CbG) | 9 | 13 | 7 | 
| Gene interacts Gene (GiG) | 9 | 20 | 8 | 

> Table 2. The distribution of each label function per relationship. 

Reviewer 2: use tablenos
